riscv/src/lib.rs

```diff
@@ -195,45 +200,105 @@ runtime = {{ path = "./runtime" }}
     )
     .unwrap();
 
-    compile_rust_crate_to_riscv_asm(cargo_file.to_str().unwrap())
+    compile_rust_crate_to_riscv_asm(cargo_file.to_str().unwrap(), output_dir)
 }
 
-pub fn compile_rust_crate_to_riscv_asm(input_dir: &str) -> BTreeMap<String, String> {
-    let temp_dir = Temp::new_dir().unwrap();
+macro_rules! as_ref [
+    ($t:ty; $($x:expr),* $(,)?) => {
+        [$(AsRef::<$t>::as_ref(&$x)),+]
+    };
+];
+
+pub fn compile_rust_crate_to_riscv_asm(
+    input_dir: &str,
+    output_dir: &Path,
+) -> BTreeMap<String, String> {
+    let target_dir = output_dir.join("cargo_target");
 
-    let cargo_status = Command::new("cargo")
+    // We call cargo twice, once to get the build plan json, so we know exactly
+    // which object file to use, and once to perform the actual building.
+    let args = as_ref![
+        OsStr;
+        "+nightly-2023-01-03",
+        "build",
+        "--release",
+        "-Z",
+        "build-std=core,alloc",
+        "--target",
+        "riscv32imc-unknown-none-elf",
+        "--lib",
+        "--target-dir",
+        target_dir,
+        "--manifest-path",
+        input_dir,
+        // These 3 arguments must come last, as they will be removed:
+        "-Z",
+        "unstable-options",
+        "--build-plan"
+    ];
+
+    let build_status = Command::new("cargo")
         .env("RUSTFLAGS", "--emit=asm -g")
-        .args([
-            "+nightly-2023-01-03",
-            "build",
-            "--release",
-            "-Z",
-            "build-std=core,alloc",
-            "--target",
-            "riscv32imc-unknown-none-elf",
-            "--lib",
-            "--target-dir",
-            temp_dir.to_str().unwrap(),
-            "--manifest-path",
-            input_dir,
-        ])
+        .args(&args[0..(args.len() - 3)])
```

Comment:
The `-3` is kind of magic. Can't you store those three in a separate array and concatenate them or something?

Reply:
fixed